sandbox/apparmor/apparmor.go

```diff
@@ -439,6 +441,32 @@ func PromptingSupportedByFeatures(apparmorFeatures *FeaturesSupported) (bool, st
 	if !strutil.ListContains(apparmorFeatures.ParserFeatures, "prompt") {
 		return false, "apparmor parser does not support the prompt qualifier"
 	}
+	// In the future, the complain redirect will be added to apparmor in the
+	// kernel, and it will share the same notify framework as the listener, so
+	// the presence of the notification socket will no longer be sufficient to
+	// indicate that the system supports prompting (assuming the kernel and
+	// parser features checks above pass). When this lands, an additional
+	// directory will be added at featuresSysPath/policy/notify. At this point,
+	// there will be a file at featuresSysPath/policy/notify/user which contains
+	// a list of mediation classes for which prompting is supported. Thus:
+	// - If featuresSysPath/policy/notify does not exist, and the above checks
+	//   pass, then prompting is supported.
+	// - If featuresSysPath/policy/notify exists, but does not contain a file
+	//   called user, then prompting is not supported.
+	// - If featuresSysPath/policy/notify exists and includes a file called
+	//   user, then prompting is supported.
+	// The user file should contain a list of mediation classes which are
+	// supported, and this should at the bare minimum include "file", so check
+	// for its presence among the kernel features.
+	if strutil.ListContains(apparmorFeatures.KernelFeatures, "policy:notify") {
+		if !strutil.ListContains(apparmorFeatures.KernelFeatures, "policy:notify:user:file") {
+			return false, "the kernel does not support prompting for file access"
+			// XXX: should this error message be "apparmor kernel features do not support prompting" as well?
+		}
+	}
+	if !notify.SupportAvailable() {
+		return false, "kernel notification socket required by listener is not present"
```

Comment:
this message doesn't make clear that this issue related to prompting, in general I agree with Alex that some kind of common starting prefix/prefix would be good

Reply:
How does this sound?

```
--- a/sandbox/apparmor/apparmor.go
+++ b/sandbox/apparmor/apparmor.go
@@ -460,12 +460,11 @@ func PromptingSupportedByFeatures(apparmorFeatures *FeaturesSupported) (bool, st
        // for its presence among the kernel features.
        if strutil.ListContains(apparmorFeatures.KernelFeatures, "policy:notify") {
                if !strutil.ListContains(apparmorFeatures.KernelFeatures, "policy:notify:user:file") {
-                       return false, "the kernel does not support prompting for file access"
-                       // XXX: should this error message be "apparmor kernel features do not support prompting" as well?
+                       return false, "apparmor kernel features do not support prompting for file access"
                }
        }
        if !notify.SupportAvailable() {
-               return false, "kernel notification socket required by listener is not present"
+               return false, "apparmor kernel notification socket required by prompting listener is not present"
        }
        return true, ""
 }
```

Since these are "unsupported reason" strings returned by the `apparmor.PromptingSupported` function, I don't think a full-on prefix like `apparmor prompting: ...` is necessary, but I do think having the common `apparmor kernel ...` and including the word `prompting` later could at least make things a bit clearer?